### ISSUE TITLE
Clarification: current `cocoa.py` seems to differ from the COCOA algorithm described in the paper

### ISSUE THREAD

**Issue by keith-Jiang** [infusion-zero-edit/CoCoA#2]
Thanks for open-sourcing this work.                                                                     
                                                                                                              
  I was comparing the paper with the current `cocoa.py` implementation, and I’m confused about whether the    
  repo is intended to reproduce the published COCOA algorithm exactly, or a later heuristic variant.          
                                                                                                              
  From the paper, my understanding is:                                                                        
                                                                                                              
  - COCOA defines a blended distribution                                                                      
    `q(y_t) ∝ p(y_t)^(1-t) * p_ctx(y_t)^t`
    with renormalization (Sec. 3.3 / Eq. 3).
  - The interpolation weight `t` is adaptive, based on conflict score / Rényi divergence / entropy gap /      
  contextual peakedness (Eqs. 9–10).                                                                          
  - The paper also seems to treat fixed gating as an ablation rather than the main method (`w/o Adaptive      
  Gating (=0.5)`).                                                                                            
                                                                                                              
  But in the current code I noticed a few things that seem different:                                         
                                                                                                              
  1. `global_alpha` is computed adaptively and then immediately overwritten by a constant:        

     ```python                                                                                                
     raw_a = torch.sigmoid(...)                                                                               
     global_alpha = torch.clamp(raw_a, 0.1, 0.9)                                                              
     global_alpha = 0.5                                                                                       
  2. The final score does not look like a simple power interpolation / log-prob blend anymore. It includes    
     extra terms such as:                                                                                     
      - gamma * Delta                                                                                         
      - history penalty                                                                                       
      - margin/z-score peak amplification                                                                     
      - uniform smoothing / floor clipping                                                                    
      - EOS lock                                                                                              
      - top-p projection sampling                                                                             
  3. The divergence computation in code also does not obviously match the paper’s Rényi-divergence formula,   
     although I may be misunderstanding that part.                                                            
                                                                                                              
  So I’m not sure whether:                                                                                    
                                                                                                              
  - cocoa.py is the exact implementation used for the paper results, or                                       
  - it is an experimental extension of COCOA.                                                                 
                                                                                                              
  Could you clarify this?                                                                                     
                                                                                                              
  If the current code is a variant, it might help reproducibility to either:                                  
                                                                                                              
  - provide a minimal paper-faithful implementation, or                                                       
  - document clearly which parts are post-paper heuristics.
                                                                                                              
  Thanks.
